core/src/Cabana_AoSoA.hpp

```diff
@@ -379,6 +396,42 @@ class AoSoA
         _data = resized_data;
     }
 
+    /*!
+      \brief Remove unused capacity.
+
+      Will reduce the capacity to be the smallest number of SoAs needed to
+      hold size() tuples. If reallocation occurs, all slices and all
+      references to the elements are invalidated. If no reallocation takes
+      place, no slices or references are invalidated.
+    */
+    void shrinkToFit()
+    {
+        static_assert( !memory_traits::is_unmanaged,
+                       "Cannot shrink unmanaged memory" );
+
+        // If we aren't asking for any fewer SoA objects then we have nothing
+        // to do. The amount of allocated data has to be at least as big as
+        // _num_soa so we just need to check here that they are equivalent. If
+        // they are equivalent, the container is already as small as it can be.
+        if ( _data.size() == _num_soa )
+            return;
+
+        // Assign the new capacity.
+        _capacity = _num_soa * vector_length;
+
+        // We need fewer SoA objects so allocate a new view and copy the
+        // existing data.
+        soa_view resized_data(
+            Kokkos::ViewAllocateWithoutInitializing( _data.label() ),
+            _num_soa );
+        if ( _num_soa > 0 )
+            Kokkos::deep_copy(
+                resized_data,
+                Kokkos::subview( _data, Kokkos::pair<size_type, size_type>(
+                                            0, _num_soa ) ) );
+        _data = resized_data;
+    }
```

Comment:
Per an offline discussion with Stuart, this will obviously cost ~2x in memory (possibly blowing out and causing an OOM), as well as have some performance penalty to copying the data over. Users may expect something like a shrink to be cheap and get shot in the foot. As things stands (wrt the kokkos resize initialization stuff), there is no good way to avoid this, but it might be nice to somehow communicate it to potential users?

Eventually it might be nice to rely on `Kokkos::resize` of our underlying views, but this is good enough for now